docs/designs/011-kubernetes-object-monitor.md

```diff
@@ -0,0 +1,283 @@
+Time-based conditions can be expressed directly in CEL using built-in timestamps:
+- Node conditions: `(now - timestamp(resource.status.conditions[0].lastTransitionTime)) > duration('2h')`
+- Event timestamps: `(now - timestamp(resource.eventTime)) < duration('5m')` (for recent events only)
+
+
+#### CEL Variables
+
+- `resource` - The Kubernetes resource object being evaluated (type depends on policy's GVK)
+  - For Node policies: `resource` is `corev1.Node`
+  - For Event policies: `resource` is `eventsv1.Event`
+- `now` - Current timestamp (time.Time)
+
+### State Machine
+
+Uniform two-state machine for all resources:
+
+```text
+┌─────────────┐
+│  Unmatched  │
+└──────┬──────┘
+       │ Predicate evaluates to true
+       ↓
+┌─────────────┐
+│   Matched   │──→ Send Unhealthy Event
+└──────┬──────┘
+       │ Predicate evaluates to false
+       ↓
+┌─────────────┐
+│  Unmatched  │──→ Send Healthy Event
+└─────────────┘
+```
+
+#### Key Characteristics
+
+- Predicate evaluates to `true` → Send unhealthy health event
+- Predicate evaluates to `false` → Send healthy/recovery health event
+- Time conditions expressed directly in CEL (e.g., `now - lastTransitionTime > 2h`)
+- No state persistence needed - all logic in CEL predicates
+
+
+### Reconciliation Triggers
+
+The controller reconciles on two triggers:
+
+#### 1. Resource Changes (Event-Driven)
+
+- Informers watch configured GVK resources (e.g., Nodes, Events)
+- When a watched resource changes (create/update/delete), triggers reconciliation
+- For each policy matching the resource's GVK, evaluate the CEL predicate
+
+#### 2. Periodic Reconciliation
+
+- Required for time-based predicates (e.g., "NotReady for more than 2 hours")
+- When a Node condition transitions to NotReady, the predicate evaluates to `false` initially
+- After 2 hours, periodic reconciliation re-evaluates and detects the threshold crossing
+- Configured via controller's resync period (e.g., every 5 minutes)
+
+#### Evaluation Result
+
+- Predicate evaluates to `true` → Send unhealthy event
+- Predicate evaluates to `false` → Send healthy event
+
+### Reconciliation Logic
+
+Uniform reconciliation for all resources:
+
+1. Retrieve the resource object (from informer cache)
+2. For each policy matching the resource's GVK, evaluate the CEL predicate
+3. If predicate evaluates to `true`:
+   - Publish unhealthy HealthEvent via gRPC
+4. If predicate evaluates to `false`:
+   - Publish healthy/recovery HealthEvent via gRPC
+5. Multiple reconciliations run concurrently (configured via `MaxConcurrentReconciles`)
+
+#### Special Handling for Pod Events
+
+For Event resources where `resource.regarding.kind == 'Pod'`:
+- Look up the Pod to determine which Node it's scheduled on
+- Associate the health event with that Node in the HealthEvent payload
+
+
+
+
+### Recovery Event Handling
+
+Recovery is handled uniformly for all resource types:
+
+- When CEL predicate evaluates to `false`, a healthy/recovery HealthEvent is published
```

Comment:
How would this work for ephemeral objects like events? Events are deleted after a TTL so we need to have some way to track that an event that caused an unhealthy event doesn't exist anymore. Maybe we maintain an in-memory list of ephemeral objects that have triggered an unhealthy event and we send a healthy event when the event is deleted, or it has not been seen after X min. The state doesn't have to be written to disk because we should be able to reconstruct the state on the initial re-list.

We can have a configuration of something to the effect of `isEphemeral` and then based on that add the required logic.

Reply:
Hmm, good point. I'll take back `or it has not been seen after X min`. For events specifically though since they get GC'd after an hour, `or it has not been seen after X min` would end up being auto deleted after 1 hour.

So, can we do:
1. on a cold start, if nothing matches send a healthy event
2. on regular operation, on delete of an object we are tracking, send healthy event